[text-wrap-style] Add faster algorithm for pretty.

yulunwu2 opened the pull request:
#### dbe441a76e45df563e5756de4721114bca3a4920
<pre>
[text-wrap-style] Add faster algorithm for pretty.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287596">https://bugs.webkit.org/show_bug.cgi?id=287596</a>
&lt;<a href="https://rdar.apple.com/144742121">rdar://144742121</a>&gt;

Reviewed by Alan Baradlay.

[text-wrap-style:pretty] is expected to be more performant than
text-wrap-style:balance as it is intended for use on larger blocks
of text.

However, the current implementation is overly complex and
significantly slower. We should update text-wrap-style:pretty to
be more performant by removing multi solution tracking, as well
as adding minimum/maximum line width heuristics per Knuth-Plass.
This simultaneously reduces search space as well as acts as a
trigger for hyphenation.

Hyphenation support will be landed in a follow up PR.

Canonical link: <a href="https://commits.webkit.org/290568@main">https://commits.webkit.org/290568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc7896659a42029cb51f0cb027c6c0c024387b07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45335 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95414 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69603 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27174 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93409 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82008 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49952 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7640 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36367 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40320 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77975 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97246 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17596 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78594 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17853 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77810 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19220 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22258 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20863 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10865 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17606 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22942 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17347 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20799 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19131 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -172,6 +183,13 @@ void InlineContentConstrainer::checkCanConstrainInlineItems()
     }
 }
 
+bool InlineContentConstrainer::panicHypenate()
```
I would rename this to something that better reflects what this function actually does.

yulunwu2 replied:
Done.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -488,27 +506,15 @@ std::optional<Vector<LayoutUnit>> InlineContentConstrainer::prettifyRange(Inline
-    auto firstLineTextIndent = m_inlineFormattingContext.formattingUtils().computedTextIndent(InlineFormattingUtils::IsIntrinsicWidthMode::No, previousLineEndsWithLineBreak, m_maximumLineWidth);
-    auto textIndent = m_inlineFormattingContext.formattingUtils().computedTextIndent(InlineFormattingUtils::IsIntrinsicWidthMode::No, false, m_maximumLineWidth);
+    auto firstLineTextIndent = computeTextIndent(previousLineEndsWithLineBreak);
+    auto textIndent = computeTextIndent(false);
```
looks like boolean trap.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -520,8 +526,19 @@ std::optional<Vector<LayoutUnit>> InlineContentConstrainer::prettifyRange(Inline
-        recordAndMaintainBestSolutions(breakIndex, EntryPretty { .accumulatedCost = cost, .previousBreakIndex = 0, .lastLineWidth = firstLineCandidateWidth });
+        auto cost = computeCostPretty(firstLineCandidateWidth, idealLineWidth, breakIndex, numberOfBreakOpportunities, idealLineWidth);
+        if (cost < std::numeric_limits<float>::infinity())
+            lastValidBreakpoint = fmax(lastValidBreakpoint, breakIndex);
```
fmax vs. size_t?

yulunwu2 replied:
Done.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -65,13 +64,20 @@ struct EntryPretty {
-    auto intermediate = difference > 0 ? difference / textWrapPrettyShrinkability:
-    difference / textWrapPrettyStretchability;
+    auto difference = candidateLineWidth - idealLineWidth;
+    auto intermediate = difference > 0 ? difference / textWrapPrettyStretchability: difference / textWrapPrettyShrinkability;
```
textWrapPrettyStretchability :

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -520,8 +526,19 @@ std::optional<Vector<LayoutUnit>> InlineContentConstrainer::prettifyRange(Inline
+    if (!lastValidBreakpoint) {
+        // If hyphenation does not create a valid solution, we should return early.
+        if (!panicHypenate())
+            return std::nullopt;
```
return { }

yulunwu2 replied:
Done here and else where.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -172,6 +183,15 @@ void InlineContentConstrainer::checkCanConstrainInlineItems()
+    // TODO: Split the hyphenated inlineItem into two inline text items.
+    // TODO: Return the next valid breakpoint.
+    // TODO: This will also change numberOfBreakOpportunities and we should update state[].
+    return std::optional<size_t> { };
```
return { } would just do.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.h:
```diff
@@ -36,6 +36,11 @@
 namespace WebCore {
 namespace Layout {
 
+enum class PreviousLineBreakStatus {
```
any reason why it is not private to `InlineContentConstrainer`?

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.h:
```diff
@@ -36,6 +36,11 @@
 
+enum class PreviousLineBreakStatus {
+    EndedWithLineBreak,
+    DidNotEndWithLineBreak
```
are you planning on expanding this enum? If not you could just use `enum class PreviousLineEndsWithLineBreak : bool { No, Yes };`

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.h:
```diff
@@ -43,10 +43,16 @@ class InlineContentConstrainer {
 private:
     friend struct SlidingWidth;
+    enum class PreviousLineEndsWithLineBreak {
+        None,
```
What does 'none' mean in this context?  No previous line? (in which case, I think it should be optional<>)

yulunwu2 committed:
[text-wrap-style] Add faster algorithm for pretty.
https://bugs.webkit.org/show_bug.cgi?id=287596
<rdar://144742121>

Reviewed by Alan Baradlay.

[text-wrap-style:pretty] is expected to be more performant than
text-wrap-style:balance as it is intended for use on larger blocks
of text.

However, the current implementation is overly complex and
significantly slower. We should update text-wrap-style:pretty to
be more performant by removing multi solution tracking, as well
as adding minimum/maximum line width heuristics per Knuth-Plass.
This simultaneously reduces search space as well as acts as a
trigger for hyphenation.

Hyphenation support will be landed in a follow up PR.

Canonical link: https://commits.webkit.org/290568@main
